drivers: adc: nrfx_saadc: enable negative values for single-ended ADC readings

jaz1-nordic opened the pull request:
The ADC driver API already supports ADC readings which can return signed values, these are differential readings. In Nordic's datasheet, we have a mode called "single ended", but its just a name. "Single ended" is a differential reading, with the negative channel tied to GND. This is not compatible with zephyrs definition of a single ended reading.

To support Nordic's "single ended" mode, the user must configure a differential reading, with the negative input tied to ground, which the saadc driver can then use to configure the reading as Nordic SAADC "single ended", and return negative values as expected.

bjarki-andreasen reviewed (dismissed):
This looks pretty much perfect

anangl commented on include/zephyr/dt-bindings/adc/nrf-saadc.h:
```diff
@@ -7,6 +7,7 @@
 #ifndef ZEPHYR_INCLUDE_DT_BINDINGS_ADC_NRF_SAADC_H_
 #define ZEPHYR_INCLUDE_DT_BINDINGS_ADC_NRF_SAADC_H_
 
+#define NRF_SAADC_GND  0
```
It should be mentioned here that this value is only applicable for `negative-input` and it will cause the sampling to be performed in single-ended mode, which for SAADC is in fact differential but with the negative input shorted to ground internally, but in this case negative values will not be rounded to 0.

bjarki-andreasen replied:
while also, making it very clear "single-ended" here is not zephyr's "single-ended", its nordic SAADC "single-ended" which is quite different in "contract" :) Its a tricky thing to communicate right

bjarki-andreasen replied:
I would suggest something like:

```
/**
 * @brief Short ADC negative input to ground
 * 
 * @details The nRF SAADC hardware only supports differential readings.
 * The nRF SAADC SE (single ended) mode is differential with the negative
 * input shorted to GND (ground). To use the nRF SAADC SE mode, set the
 * negative input to NRF_SAADC_GND:
 *
 *   zephyr,input-positive = <NRF_SAADC_AIN3>;
 *   zephyr,input-negative = <NRF_SAADC_GND>;
 *
 * The nRF SAADC driver also supports using the nRF SAADC SE mode in
 * emulated "single-ended" mode, as defined by zephyr. In this mode,
 * negative readings will be clamped to 0 by software, to emulate the
 * behavior of an ADC in "single-ended" mode, as defined by zephyr. To
 * do this, only define the positive input:
 *
 *   zephyr,input-positive = <NRF_SAADC_AIN3>;
 */
```

jaz1-nordic commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -366,8 +369,14 @@ static int adc_nrfx_channel_setup(const struct device *dev,
 	 */
 	if (channel_cfg->differential) {
-		ch_cfg->mode = NRF_SAADC_MODE_DIFFERENTIAL;
-		m_data.single_ended_channels &= ~BIT(channel_cfg->channel_id);
+		if (channel_cfg->input_negative == NRF_SAADC_GND) {
+			ch_cfg->mode = NRF_SAADC_MODE_SINGLE_ENDED;
```
The result will be correct because we're not changing the mode. We don't want to use Diff mode here, but single-ended. This change is only intended to prevent negative values from being corrected for a given channel (which will still operate as single-ended). The result remains unchanged. We're "artificially" configuring the Diff mode by configuring the second input, but the actual mode remains single-ended.

jaz1-nordic replied:
@anangl please take a look and remove your change-request.

nika-nordic reviewed (dismissed):
LGTM - no vendor-specific API means less maintenance in the future

bjarki-andreasen reviewed (dismissed):
Cool :)

anangl commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -539,11 +549,18 @@ static void correct_single_ended(const struct adc_sequence *sequence, nrf_saadc_
-		for (int i = 0; i < num_samples; i++) {
-			if (sample[i] < 0) {
-				sample[i] = 0;
+		if (divide_single_ended_value) {
```
```suggestion
		if (sequence->channels & divide_single_ended_value) {
```

jaz1-nordic committed:
drivers: adc: nrfx: enable negative values for single-ended ADC readings

The ADC driver API already supports ADC readings which can return signed
values, these are differential readings. In Nordic's datasheet, we have
a mode called "single ended", but its just a name. "Single ended" is a
differential reading, with the negative channel tied to GND. This is not
compatible with zephyrs definition of a single ended reading.

To support Nordic's "single ended" mode, the user must configure
a differential reading, with the negative input tied to ground, which
the saadc driver can then use to configure the reading as Nordic SAADC
"single ended", and return negative values as expected.

Signed-off-by: Jakub Zymelka <jakub.zymelka@nordicsemi.no>
